### PR TITLE
[BH-1195] Fix audio playback while volume change

### PR DESCRIPTION
**Pull request by tigransoghbatyan-digica**
Prevent audio restart on volume setting change

### CODE REVIEW

File: products/BellHybrid/apps/application-bell-settings/presenter/alarm_settings/PrewakeUpPresenter.cpp [hunk 1]
```diff
@@ -30,7 +30,9 @@ namespace app::bell_settings
         this->provider->onVolumeExit   = [this](const auto &) { this->stopSound(); };
         this->provider->onVolumeChange = [this, playSound](const auto &val) {
             this->audioModel->setVolume(val, AbstractAudioModel::PlaybackType::PreWakeup, {});
-            playSound(currentSoundPath);
+            if (this->audioModel->hasPlaybackFinished()) {
```
Comment (mjan-dev) [inline products/BellHybrid/apps/application-bell-settings/presenter/alarm_settings/PrewakeUpPresenter.cpp:33]: why only the prewakeup presenter checks if the playback has finished?

Comment (tigransoghbatyan-digica) [reply products/BellHybrid/apps/application-bell-settings/presenter/alarm_settings/PrewakeUpPresenter.cpp:33]: fixed

File: products/BellHybrid/apps/include/AudioApplication.hpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+                         StartInBackground startInBackground = {false},
+                         uint32_t stackDepth                 = 4096,
+                         sys::ServicePriority priority       = sys::ServicePriority::Idle)
+            : Application(std::move(name), std::move(parent), statusIndicators, startInBackground, stackDepth),
+              audioModel{std::make_shared<AudioModel>(this)}
+        {
+            bus.channels.push_back(sys::BusChannel::ServiceAudioNotifications);
+        }
+    };
+} // namespace app
```
Comment (swichu91) [inline products/BellHybrid/apps/include/AudioApplication.hpp:30]: I get you are trying to remove unnecessary code duplication, which is fine but I'm not sure, if this is the correct solution. I strongly believe, the composition should be used in favor of inheritance. After all, it's just two lines more that needs to be added to applications:
```
std::unique_ptr<AbstractAudioModel> audioModel;
bus.channels.push_back(sys::BusChannel::ServiceAudioNotifications);
```

File: products/BellHybrid/apps/include/AudioApplication.hpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#include <Application.hpp>
+#include <common/models/AbstractAudioModel.hpp>
+#include <common/models/AudioModel.hpp>
+
+namespace app
+{
+    class AudioApplication : public Application
+    {
+      protected:
+        std::shared_ptr<AbstractAudioModel> audioModel;
```
Comment (swichu91) [inline products/BellHybrid/apps/include/AudioApplication.hpp:15]: Do we really need a shared pointer here?

File: products/BellHybrid/apps/common/include/common/models/BedtimeModel.hpp [hunk 4]
```diff
@@ -65,7 +65,7 @@ namespace app::bell_bedtime
       public:
         BedtimeModel() = delete;
 
-        explicit BedtimeModel(ApplicationCommon *app, std::unique_ptr<AudioModel> &&audioModel)
+        explicit BedtimeModel(ApplicationCommon *app, std::shared_ptr<AbstractAudioModel> audioModel)
```
Comment (swichu91) [inline products/BellHybrid/apps/common/include/common/models/BedtimeModel.hpp:68]: Does `BedTimeModel` really need to have a shared pointer to `audioModel`? Shared pointers are much slower than unique pointers or references. Moreover, the cases where you need them are rather rare. In this case, a simple reference will be sufficient. This also applies to other models/presenters.

**Review by swichu91 — APPROVED** [review]

**Review by mjan-dev — APPROVED** [review]

**Comment by tigransoghbatyan-digica** [comment]
Merged in aggregated PR

**Pull request closed without merging**